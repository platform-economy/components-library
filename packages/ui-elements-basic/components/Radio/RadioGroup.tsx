import * as React from 'react';

export interface RadioGroupContextInterface{
  onValueChange?: (value: string | string[] | number, ...args: unknown[]) => void;
  value: string | string[] | number;
}

export interface RadioGroupProps extends RadioGroupContextInterface {
  groupName?:string;
}

export interface RadioGroupState {
  radioGroupContext:RadioGroupContextInterface;
}

export const RadioGroupContext =
  React.createContext<RadioGroupContextInterface|null>(null);

export class RadioGroup extends  React.Component<RadioGroupProps, RadioGroupState>{

  constructor(props:RadioGroupProps, ...args:unknown[]) {
    super(props, ...args);
    this.handleValueChange = this.handleValueChange.bind(this);
    this.state = {
      radioGroupContext: {
        value: props.value,
        onValueChange: this.handleValueChange,
      },
    };
  }

  handleValueChange(value: string | string[] | number) {
    if (this.props.onValueChange != null) {
      this.props.onValueChange(value);
    }
  }

  static getDerivedStateFromProps(newProps: RadioGroupProps, state: RadioGroupState) {
    if (newProps.value !== state.radioGroupContext.value) {
      return {
        radioGroupContext: {
          ...state.radioGroupContext,
          value: newProps.value,
        },
      };
    }
    return null;
  }

  render() {
    return(
      <RadioGroupContext.Provider value={this.state.radioGroupContext}>
        {this.props.children}
      </RadioGroupContext.Provider>
    );
  }
}
