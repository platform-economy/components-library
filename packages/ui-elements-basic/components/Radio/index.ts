export * from './Radio';
export * from './RadioGroup';
export * from './RadioToggle';
