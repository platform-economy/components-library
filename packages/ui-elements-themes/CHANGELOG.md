# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.2](https://github.com/relayr/ui-elements/compare/@relayr/ui-elements-themes@0.1.1...@relayr/ui-elements-themes@0.1.2) (2019-08-22)

**Note:** Version bump only for package @relayr/ui-elements-themes






## [0.1.1](https://github.com/relayr/ui-elements/compare/@relayr/ui-elements-themes@0.1.0...@relayr/ui-elements-themes@0.1.1) (2019-08-21)

**Note:** Version bump only for package @relayr/ui-elements-themes






# 0.1.0 (2019-08-02)


### Features

* **ui-elements-basic:** new InputLabel component ([abb6076](https://github.com/relayr/ui-elements/commit/abb6076))
* **ui-elements-layout:** add Card component ([e8db0cf](https://github.com/relayr/ui-elements/commit/e8db0cf))
* **ui-elements-tables:** Table components ([#102](https://github.com/relayr/ui-elements/issues/102)) ([b1b8d6f](https://github.com/relayr/ui-elements/commit/b1b8d6f))
* **ui-elements-themes:** custom font linking system ([8e9f0f0](https://github.com/relayr/ui-elements/commit/8e9f0f0))
