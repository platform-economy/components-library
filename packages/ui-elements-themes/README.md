# ui-elements-themes

## Add to project

via npm:

```bash
npm install --save @relayr/ui-elements-themes
```

or Yarn:
```bash
yarn add @relayr/ui-elements-themes
```
