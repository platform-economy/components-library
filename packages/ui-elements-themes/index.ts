export * from './themes/light';
export * from './themes/dark';
export * from './themes/test';
export * from './themes/_common/paletteDark';
export * from './themes/_common/paletteLight';
export * from './themes/_common/methods';
export * from './themes/_common/fonts';
export * from './themes/_common/withTestTheme';
