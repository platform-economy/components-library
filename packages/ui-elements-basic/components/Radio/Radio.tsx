import * as React from 'react';
import styledComponents from 'styled-components';
import { RadioGroupContext, RadioGroupContextInterface } from './RadioGroup';
import classes from 'classnames';

export interface RadioProps extends React.InputHTMLAttributes<HTMLInputElement> {
  groupName?:string;
  checked?:boolean;
  value: string | string[] | number;
  onValueChange?: (value: string | string[] | number, ...args: unknown[]) => void;
}

interface RadioState {
  hasFocus:boolean;
}

const StyledInputWrapper = styledComponents.span<React.HTMLAttributes<HTMLSpanElement>>`
  & > input[type=radio]{
    opacity: 0;
    display: inline;
    width:0px;
    height:0px;
    margin:0px;
    padding:0px;
  }
  & svg {
    margin: 0px 5px
    width: 1em;
    height: 1em;
  }
  & svg .radio-button-border{
    stroke-width: 41;
    stroke: ${props => props.theme.palette.text_inverted};
    fill: transparent;
  }
  &.focus svg .radio-button-border{
    stroke: ${props => props.theme.palette.active};
  }
  &:hover svg .radio-button-border,
  &:focus svg .radio-button-border{
    stroke: ${props => props.theme.palette.text};
  }
  &:active svg .radio-button-border{
    stroke: ${props => props.theme.palette.active_lowered};
  }
  &:disabled svg .radio-button-border,
  &.disabled svg .radio-button-border{
    stroke: ${props => props.theme.palette.inactive};
  }

  & svg .radio-button-selection{
    fill: ${props => props.theme.palette.text_inverted};
  }
  &.focus svg .radio-button-selection{
    fill: ${props => props.theme.palette.active};
  }
  &:hover svg .radio-button-selection,
  &:focus svg .radio-button-selection{
    fill: ${props => props.theme.palette.text};
  }
  &:active svg .radio-button-selection{
    fill: ${props => props.theme.palette.active_lowered};
  }
  &:disabled svg .radio-button-selection,
  &.disabled svg .radio-button-selection{
    fill: ${props => props.theme.palette.inactive};
  }
  &.checked svg .radio-button-selection{
    display:auto;
  }
  &.unchecked svg .radio-button-selection{
    display:none;
  }
  &.disabled label{
    color: ${props => props.theme.palette.inactive};
  }
`;

export class Radio extends React.Component<RadioProps, RadioState> {
  static contextType = RadioGroupContext;
  context:RadioGroupContextInterface;
  private inputElement: React.RefObject<HTMLInputElement> = React.createRef<HTMLInputElement>();

  constructor(props: Readonly<RadioProps>) {
    super(props);
    this.state = { hasFocus: false };
    this.internalOnClick = this.internalOnClick.bind(this);
    this.internalOnBlur = this.internalOnBlur.bind(this);
    this.internalOnFocus = this.internalOnFocus.bind(this);
    this.internalOnChange = this.internalOnChange.bind(this);
  }

  private internalOnClick(event: React.MouseEvent<HTMLInputElement>) {
    if (this.inputElement.current !== event.target) {
      // redirect the click event to pass throw radio input
      this.inputElement.current!.focus();
      this.inputElement.current!.click();
      event.stopPropagation();
    } else {
      if (this.props.onClick != null) {
        this.props.onClick(event);
      }
    }
  }
  private internalOnFocus(event: React.FocusEvent<HTMLInputElement>) {
    this.setState({ hasFocus: true });
    if (this.props.onFocus) {
      this.props.onFocus(event);
    }
  }
  private internalOnBlur(event: React.FocusEvent<HTMLInputElement>) {
    this.setState({ hasFocus: false });
    if (this.props.onBlur) {
      this.props.onBlur(event);
    }
  }
  private internalOnChange(event: React.ChangeEvent<HTMLInputElement>) {
    if (this.props.onValueChange != null && event.target.checked) {
      this.props.onValueChange(this.props.value);
    }
    if (this.props.onChange != null) {
      this.props.onChange(event);
    }
    if (this.context &&
      this.context.onValueChange &&
      event.target.checked) {
      this.context.onValueChange(this.props.value);
    }
  }

  get isChecked():boolean {
    if (this.context) {
      return this.context.value === this.props.value;
    }
    return this.props.checked || false;
  }

  get hasFocus():boolean {
    return this.state.hasFocus;
  }

  render() {
    const {
      groupName,
      checked,
      children,
      value,
      onClick,
      onValueChange,
      onChange,
      onFocus,
      onBlur,
      ...inputProps } = this.props;
    return (
      <StyledInputWrapper
        onClick={this.internalOnClick}
        className={classes(
          {
            checked: this.isChecked,
            unchecked: !this.isChecked,
            disabled: inputProps.disabled,
            focus: this.hasFocus,
          })}
        {...inputProps}
      >
        <input
          type="radio"
          name={this.props.groupName}
          value={value}
          onFocus={this.internalOnFocus}
          onBlur={this.internalOnBlur}
          onChange={this.internalOnChange}
          ref={this.inputElement}
          disabled={inputProps.disabled}
        />
        <svg viewBox="0 0 512 512">
          <circle className="radio-button-border" cx="256" cy="328" r="160"/>
          <circle className="radio-button-selection" cx="256" cy="328" r="80"/>
        </svg>
        {children}
      </StyledInputWrapper>
    );

  }
}
