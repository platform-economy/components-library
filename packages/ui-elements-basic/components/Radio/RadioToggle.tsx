import * as React from 'react';
import { ToggleProps } from '../Toggle/types';
import { RadioGroupContext, RadioGroupContextInterface } from './RadioGroup';
import { Toggle } from '../Toggle';

export class RadioToggle extends  React.Component<ToggleProps> {
  static contextType = RadioGroupContext;
  context:RadioGroupContextInterface;

  constructor(props: Readonly<ToggleProps>) {
    super(props);
    this.internalOnClick = this.internalOnClick.bind(this);
  }

  private internalOnClick(event: React.MouseEvent<HTMLButtonElement>) {
    if (this.props.onClick != null) {
      this.props.onClick(event);
    }

    if (this.context &&
      this.context.onValueChange &&
      this.props.value !== undefined) {
      this.context.onValueChange(this.props.value);
    }
  }

  get isToggled():boolean {
    if (this.props.hasOwnProperty('toggled')) {
      return !!this.props.toggled;
    }
    if (this.context) {
      return this.context.value === this.props.value;
    }
    return false;
  }

  render() {
    return (
      <Toggle {...this.props} toggled={this.isToggled} onClick={this.internalOnClick}/>
    );
  }
}
