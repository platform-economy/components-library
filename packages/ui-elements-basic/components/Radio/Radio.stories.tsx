import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { text, number, boolean, select } from '@storybook/addon-knobs';

import { Radio } from './Radio';
import { decorateAction } from '@storybook/addon-actions';
import { RadioGroup } from './RadioGroup';
import { withUncontrolledValue } from '@relayr/ui-elements-core/utils/withUncontrolledValue';
import { actionWithoutEvent } from '../../../../helpers/storybook-helpers';
import { RadioToggle } from './RadioToggle';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faPhone,
  faEnvelope,
  faAt,
} from '@fortawesome/free-solid-svg-icons';

const actionWithData = decorateAction([([data]) => [data]]);
const knobOptions = {
  range: true,
  min: .5,
  max: 10,
  step: .25,
};

storiesOf('ui-elements-basic/Radio', module)
.add('Basic Radio Input and group stateless', () => {
  const whiteColor = { color: 'white' };
  return (
    <div >
      <Radio
        groupName={text('Group Name', 'RadioButtonGroup')}
        value="1"
        onClick={actionWithData('radio-button-click')}
        onValueChange={actionWithData('radio-button-value-change')}
      >
        <span style={whiteColor}>First Option</span>
      </Radio>
      <br/>
      <Radio
        groupName={text('Group Name', 'RadioButtonGroup')}
        value="2"
        onClick={actionWithData('radio-button-click')}
        onValueChange={actionWithData('radio-button-value-change')}
      >
        <span style={whiteColor}>Second Option</span>
      </Radio>
      <br/>
      <Radio
        groupName={text('Group Name', 'RadioButtonGroup')}
        value="3"
        disabled={boolean('Disabled', false)}
        onClick={actionWithData('radio-button-click')}
        onValueChange={actionWithData('radio-button-value-change')}
      >
        <span style={whiteColor}>Disabled options</span>
      </Radio>
      <br/>
      <Radio
        style={{ fontSize: `${number('Font Size', 1, knobOptions)}em` }}
        groupName={text('Group Name', 'RadioButtonGroup')}
        value="4"
        onClick={actionWithData('radio-button-click')}
        onValueChange={actionWithData('radio-button-value-change')}
      >
        <span style={whiteColor}>Enlarge ??</span>
      </Radio>
    </div>
  )
})
.add('Basic Radio Input stateless', () => {
  const whiteColor = { color: 'white' };
  return (
    <RadioGroup
      value={select('select item', ['1', '2', '3', '4'], '1')}
      onValueChange={actionWithData('radio-button-group-value-change')}
    >
      <Radio
        groupName={text('Group Name', 'RadioButtonGroup')}
        value="1"
        onClick={actionWithData('radio-button-click')}
        onValueChange={actionWithData('radio-button-value-change')}
      >
        <span style={whiteColor}>First Option</span>
      </Radio>
      <br/>
      <Radio
        groupName={text('Group Name', 'RadioButtonGroup')}
        value="2"
        onClick={actionWithData('radio-button-click')}
        onValueChange={actionWithData('radio-button-value-change')}
      >
        <span style={whiteColor}>Second Option</span>
      </Radio>
      <br/>
      <Radio
        groupName={text('Group Name', 'RadioButtonGroup')}
        value="3"
        disabled={boolean('Disabled', false)}
        onClick={actionWithData('radio-button-click')}
        onValueChange={actionWithData('radio-button-value-change')}
      >
        <span style={whiteColor}>Disabled options</span>
      </Radio>
      <br/>
      <Radio
        style={{ fontSize: `${number('Font Size', 1, knobOptions)}em` }}
        groupName={text('Group Name', 'RadioButtonGroup')}
        value="4"
        onClick={actionWithData('radio-button-click')}
        onValueChange={actionWithData('radio-button-value-change')}
      >
        <span style={whiteColor}>Enlarge ??</span>
      </Radio>
    </RadioGroup>
  );
})
.add('Basic Radio with state', () => {
  const RadioGroupWithState = withUncontrolledValue()(RadioGroup);
  const whiteColor = { color: 'white' };
  return (
    <RadioGroupWithState
      value={select('select item', ['', '1', '2', '3'], '')}
      onValueChange={actionWithData('radio-button-group-value-change')}
    >
      <Radio
        groupName={text('Group Name', 'RadioButtonGroup')}
        value="1"
        onClick={actionWithData('radio-button-click')}
        onValueChange={actionWithData('radio-button-value-change')}
      >
        <span style={whiteColor}>First Option</span>
      </Radio>
      <br/>
      <Radio
        groupName={text('Group Name', 'RadioButtonGroup')}
        value="2"
        onClick={actionWithData('radio-button-click')}
        onValueChange={actionWithData('radio-button-value-change')}
      >
        <span style={whiteColor}>Second Option</span>
      </Radio>
      <br/>
      <span>
        <Radio
          groupName={text('Group Name', 'RadioButtonGroup')}
          value="3"
          disabled={boolean('Disabled', false)}
          onClick={actionWithData('radio-button-click')}
          onValueChange={actionWithData('radio-button-value-change')}
        >
          <span style={whiteColor}>Disabled Options</span>
        </Radio>
      </span>
    </RadioGroupWithState>
  );
})
.add('RadioToggle within RadioGroup with state', () => {
  const selectValue = select('select item', ['1', '2', '3'], '1');
  const RadioGroupWithState = withUncontrolledValue()(RadioGroup);
  return (
    <RadioGroupWithState
      value={selectValue}
      onValueChange={actionWithoutEvent('radio-button-group-value-change')}
    >
      <RadioToggle
        iconOn={<FontAwesomeIcon icon={faAt}/>}
        value={'1'}
        disabled={boolean('Disabled Email', false)}
      />
      <RadioToggle
        iconOn={<FontAwesomeIcon icon={faPhone}/>}
        value={'2'}
        disabled={boolean('Disabled Phone', false)}
      />
      <RadioToggle
        iconOn={<FontAwesomeIcon icon={faEnvelope}/>}
        value={'3'}
        disabled={boolean('Disabled Envelope', false)}
      />
    </RadioGroupWithState>
  );
});
