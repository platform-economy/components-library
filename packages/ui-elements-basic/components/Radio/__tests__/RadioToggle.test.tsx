import * as React from 'react';
import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as TestRenderer from 'react-test-renderer';
import 'jest-styled-components';

import { withTestTheme } from '@relayr/ui-elements-themes';

enzyme.configure({ adapter: new Adapter() });
/*
A set of utilities for testing Styled Components with Jest.
This package improves the snapshot testing experience and
provides a brand new matcher to make expectations on the style rules.
*/

// Component
import { RadioGroup } from '../../Radio';
import { RadioToggle } from '../RadioToggle';

describe('ui-elements-basic', () => {
  describe('Components', () => {
    describe('RadioToggle', () => {
      const handleClick = jest.fn();
      it('Should handle click event correctly', () => {
        const tree = enzyme.mount(withTestTheme(<RadioToggle onClick={handleClick} />));
        tree.find('StyledComponent').simulate('click');
        expect(handleClick).toHaveBeenCalledTimes(1);
      });
      it('renders conditional theme color based on given props', () => {
        const mockClick = () => null;
        const tree = TestRenderer
          .create(withTestTheme(
            <RadioToggle  onClick={mockClick} toggled={true} highlight={true}/>))
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
      it('Toggle Radio inside radio group', () => {
        const onValueChange = jest.fn();
        const tree = enzyme.mount(withTestTheme(
          <RadioGroup value={'toggleButtonValue'} onValueChange={onValueChange}>
            <RadioToggle
              onClick={handleClick}
              value="toggleButtonValue"
            />
            <RadioToggle
              value="otherButtonValue"
            />
          </RadioGroup>));
        tree.find('StyledComponent[value="toggleButtonValue"]').simulate('click');
        expect(onValueChange).toHaveBeenCalledTimes(1);
        expect(onValueChange).toHaveBeenCalledWith('toggleButtonValue');

        tree.find('StyledComponent[value="otherButtonValue"]').simulate('click');
        expect(onValueChange).toHaveBeenCalledTimes(2);
      });
      it('Toggle Radio un-toggled', () => {
        const tree = TestRenderer
          .create(withTestTheme(
            <RadioToggle
              toggled={false}
              iconOn={<span>ON</span>}
              iconOff={<span>OFF</span>}
            />))
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
