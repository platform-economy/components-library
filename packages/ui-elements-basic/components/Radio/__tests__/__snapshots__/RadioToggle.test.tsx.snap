// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui-elements-basic Components RadioToggle Toggle Radio un-toggled 1`] = `
.c0 {
  padding: 13px 30px;
  border-radius: 2px;
  border: none;
  outline: none;
  text-align: center;
  font-weight: bold;
  text-transform: uppercase;
  -webkit-transition: 100ms all ease-in-out;
  transition: 100ms all ease-in-out;
  background-color: transparent;
  color: #text;
  font-size: 1.5em;
  cursor: pointer;
  border-bottom: 2px solid transparent;
  margin-bottom: -2px;
}

.c0.iconOffFromIconOn {
  color: #inactive;
}

.c0:active:hover,
.c0:active:focus {
  color: #active_hovered;
}

.c0:focus,
.c0.iconOffFromIconOn:focus {
  color: #active;
  margin-bottom: -2px;
}

.c0:hover:not(:disabled),
.c0.iconOffFromIconOn:hover:not(:disabled) {
  color: #active_hovered;
}

.c0:active {
  color: #text;
}

.c0:disabled {
  color: #inactive;
  cursor: not-allowed;
}

<button
  className="c0"
  onClick={[Function]}
>
  <span>
    OFF
  </span>
</button>
`;

exports[`ui-elements-basic Components RadioToggle renders conditional theme color based on given props 1`] = `
.c0 {
  padding: 13px 30px;
  border-radius: 2px;
  border: none;
  outline: none;
  text-align: center;
  font-weight: bold;
  text-transform: uppercase;
  -webkit-transition: 100ms all ease-in-out;
  transition: 100ms all ease-in-out;
  background-color: transparent;
  color: #active;
  font-size: 1.5em;
  cursor: pointer;
  border-bottom: 2px solid transparent;
  margin-bottom: -2px;
}

.c0.iconOffFromIconOn {
  color: #inactive;
}

.c0:active:hover,
.c0:active:focus {
  color: #active_hovered;
}

.c0:focus,
.c0.iconOffFromIconOn:focus {
  color: #active;
  margin-bottom: -2px;
}

.c0:hover:not(:disabled),
.c0.iconOffFromIconOn:hover:not(:disabled) {
  color: #active_hovered;
}

.c0:active {
  color: #text;
}

.c0:disabled {
  color: #inactive;
  cursor: not-allowed;
}

<button
  className="c0"
  onClick={[Function]}
/>
`;
